src/hierarchies/mfa_hierarchy.cc

```diff
@@ -0,0 +1,262 @@
+  hypers->beta = bayesmix::to_eigen(hyperscast.beta());
+  hypers->phi = hyperscast.phi();
+  hypers->q = hyperscast.q();
+}
+
+std::shared_ptr<bayesmix::AlgorithmState::HierarchyHypers>
+MFAHierarchy::get_hypers_proto() const {
+  bayesmix::MFAPriorDistribution hypers_;
+  bayesmix::to_proto(hypers->mutilde, hypers_.mutable_mutilde());
+  bayesmix::to_proto(hypers->beta, hypers_.mutable_beta());
+  hypers_.set_alpha0(hypers->alpha0);
+  hypers_.set_phi(hypers->phi);
+  hypers_.set_q(hypers->q);
+
+  auto out = std::make_shared<bayesmix::AlgorithmState::HierarchyHypers>();
+  out->mutable_mfa_state()->CopyFrom(hypers_);
+  return out;
+}
+
+void MFAHierarchy::sample_full_cond(bool update_params) {
+  if (this->card == 0) {
+    // No posterior update possible
+    sample_prior();
+  } else {
+    sample_eta();
+    sample_mu();
+    sample_psi();
+    sample_lambda();
+  }
+}
+
+void MFAHierarchy::sample_eta() {
+  auto& rng = bayesmix::Rng::Instance().get();
+  auto llt = (Eigen::MatrixXd::Identity(hypers->q, hypers->q) +
+              state.lambda.transpose() * state.psi_inverse * state.lambda)
+                 .llt();
+  Eigen::MatrixXd sigma_eta(
+      llt.solve(Eigen::MatrixXd::Identity(hypers->q, hypers->q)));
+
+  if (state.eta.rows() != card) {
+    state.eta.resize(card, state.eta.cols());
+    state.eta = Eigen::MatrixXd::Zero(card, state.eta.cols());
+  }
+  for (size_t i = 0; i < card; i++) {
+    state.eta.row(i) = (stan::math::multi_normal_rng(
+        sigma_eta * (state.lambda.transpose()) * state.psi_inverse *
+            (data[i] - state.mu),
+        sigma_eta, rng));
+  }
+}
+
+void MFAHierarchy::sample_mu() {
+  auto& rng = bayesmix::Rng::Instance().get();
+  auto llt = (hypers->phi * Eigen::MatrixXd::Identity(dim, dim) +
+              card * state.psi_inverse)
+                 .llt();
+  Eigen::MatrixXd sigma_mu = llt.solve(Eigen::MatrixXd::Identity(dim, dim));
+
+  Eigen::VectorXd sum = Eigen::VectorXd::Zero(dim);
+
+  for (size_t i = 0; i < card; i++) {
+    Eigen::VectorXd row = state.eta.row(i);
+    sum += state.lambda * row;
+  }
+  Eigen::VectorXd mumean = sigma_mu * (hypers->phi * hypers->mutilde +
+                                       state.psi_inverse * (data_sum - sum));
+
+  state.mu = stan::math::multi_normal_rng(mumean, sigma_mu, rng);
+}
+
+void MFAHierarchy::sample_lambda() {
+  auto& rng = bayesmix::Rng::Instance().get();
+  Eigen::MatrixXd data_matrix(dim, card);
+  for (size_t i = 0; i < card; i++) {
+    data_matrix.col(i) = data[i];
+  }
+  for (size_t j = 0; j < dim; j++) {
+    auto llt = (Eigen::MatrixXd::Identity(hypers->q, hypers->q) +
+                state.eta.transpose() / state.psi[j] * state.eta)
+                   .llt();
+    Eigen::MatrixXd sigma_lambda =
+        llt.solve(Eigen::MatrixXd::Identity(hypers->q, hypers->q));
+    Eigen::VectorXd vectortemp(data_matrix.row(j));
+    state.lambda.row(j) = stan::math::multi_normal_rng(
+        sigma_lambda * state.eta.transpose() *
```

Comment:
you could save some of these multiplications in a temporary variable outside the for loop